core/src/main/java/org/elasticsearch/rest/RestRequest.java

```diff
@@ -222,4 +230,78 @@ public ByteSizeValue paramAsSize(String key, ByteSizeValue defaultValue) {
         return params;
     }
 
+    /**
+     * Does this request have content or a {@code source} parameter? Use this instead of {@link #hasContent()} if this
+     * {@linkplain RestHandler} treats the {@code source} parameter like the body content.
+     */
+    public final boolean hasContentOrSourceParam() {
+        return hasContent() || hasParam("source");
+    }
+
+    /**
+     * The {@linkplain XContentType} of either the request body or the {@code source} param if either exists and can be guessed, otherwise
+     * {@code null}. Only use this if you need to do something different if the request is plain text. Use
+     * {@link #contentOrSourceParamParser()} if you need a parser.
+     */
+    @Nullable
+    public final XContentType contentOrSourceParamXContentType() {
+        BytesReference content = contentOrSourceParam();
+        return content == null ? null : XContentFactory.xContentType(content);
```

Comment:
This is fairly rarely used so we can do what we like. For the most part they don't but we can certainly force them to.